components/butterfly/src/rumor/dat_file.rs

```diff
@@ -43,7 +44,11 @@ pub struct DatFile {
 
 impl DatFile {
     pub fn new(data_path: PathBuf) -> Result<Self> {
-        let file = File::open(&data_path).map_err(|err| Error::DatFileIO(data_path.clone(), err))?;
+        let file = OpenOptions::new().create(true)
+                                     .read(true)
+                                     .write(true)
+                                     .open(&data_path)
+                                     .map_err(|err| Error::DatFileIO(data_path.clone(), err))?;
```

Comment:
I think this is going to potentially break the atomicity of writing this file. Especially now that we potentially have multiple processes accessing it, it's important that we access this safely.

I think we could re-work this a bit more such that
1. If the supervisor calls this and there's no existing .rst file, it's created via the standard `DatFile::write` method, which will use `AtomicWriter`
1. If `rst-reader` calls this and there's no existing .rst file, it simply returns `None`/`Err`; I think an empty file vs no file is an important distinction.

Probably splitting this into two functions `DatFile::read_or_create(data_path: PathBuf, server: &Server) -> Result<Self>` and `DatFile::read() -> Option<Self>` for those two cases would make it more ergonomic. This would also eliminate the need for the `if size == 0` check in `read_into`.

I'm also a little concerned that the the abstraction here is still a bit off since we have calls like
```rust
dat_file.read_rumors::<Service>(dat_file.service_len())
```
Consumers needing to tell `dat_file` exactly the offset to read at seems like a fairly big hole in the encapsulation. This is a problem with the initial implementation, not your additions. I think it would be good to keep re-working it a bit more so we can get at something more usable.

Reply:
Previous to the addition of `rst-reader`, the only place that `DatFile::new()` was called was within the butterfly server when it started up.  Would giving `rst-reader` its own method to use, e.g. `DatFile::read()` as you suggested, solve the concerns around creating the file atomically?  I realize that doesn't provide any type of atomicity, per se, but it does resolve the issue of multiple processes trying to create the file.

I bring it up because `DatFile::write` takes `self`, which implies a certain chicken/egg problem when using it inside an initializer.  I tried changing `DatFile::write` to be a static method, but then the call site in the butterfly server ends up looking like `DatFile::write(dat_file.path(), self)`, which I dunno...feels a little odd.

Is there an ergonomic way to make this work that I'm missing?

Reply:
Nevermind, I think I have a solution to this.  Also, your point about consumers needing to tell `dat_file` the offset is taken.  That also felt off to me.  I'll update this to account for that.